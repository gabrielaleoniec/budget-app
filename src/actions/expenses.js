import database from '../firebase/firebase';

// ADD EXPENSE
export const addExpense = (expense = {}) => {
    return {
        type: "ADD_EXPENSE",
        expense
    }
}

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const date = createdAt ? new Date(createdAt).valueOf() : new Date().valueOf();

        const expense = { description, note, amount, createdAt: date };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                ...expense,
                id: ref.key
            }))
        });
    }
}

// REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id: id
});

// START REMOVE EXPENSES
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => { 
            dispatch(removeExpense({ id }))
        });
    }
}

// EDIT EXPENSE
export const updateExpense = (id, updates) => ({
    type: "UPDATE_EXPENSE",
    id,
    updates
})

// SET EXPENSES
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then(snapshot => {
            const expenses = [];

            snapshot.forEach(childSnapshot => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        }).catch((e) => {
            console.log('Error', e);
        })
    }
};
