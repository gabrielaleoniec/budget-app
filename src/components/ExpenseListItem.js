import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({id, description, note, amount, createdAt}) => {
    return <div>
        <div></div>
        <div>{description}</div>
        <div>{note}</div>
        <div>{(amount/100).toFixed(2) + ' z≈Ç'}</div>
        <div>{moment(createdAt).format('YYYY-MM-DD')}</div>
        <div><Link to={`/edit/${id}`}>Edit</Link></div>
    </div>
}

export default ExpenseListItem;