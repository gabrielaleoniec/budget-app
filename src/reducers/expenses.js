const defaultStateExpensifyReducer = [];

const expensifyReducer = (state = defaultStateExpensifyReducer, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'UPDATE_EXPENSE':
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                return expense;
            })
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state;
    }
}

export default expensifyReducer;