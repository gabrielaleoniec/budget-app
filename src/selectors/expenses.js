
const getVisibleExpenses = (expenses, { text, sortBy, sortByAsc, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const matchText = expense.description.toLowerCase().includes(text.toLowerCase());
        const matchStartDate = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const matchEndDate = typeof endDate !== 'number' || expense.createdAt <= endDate;

        return matchText && matchStartDate && matchEndDate;
    }).sort((a, b) => {
        if (sortBy === 'createdAt') {
            return sortByAsc ? b.createdAt - a.createdAt : a.createdAt - b.createdAt;
        } else if (sortBy === 'amount') {
            return sortByAsc ? b.amount - a.amount : a.amount - b.amount;
        }
    })
}

export default getVisibleExpenses;