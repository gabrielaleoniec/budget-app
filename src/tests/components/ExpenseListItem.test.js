import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

test('Should render ExpenseListItem with properties', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});