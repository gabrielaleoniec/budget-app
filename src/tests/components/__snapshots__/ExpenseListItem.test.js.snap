// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem with properties 1`] = `
<div>
  <div />
  <div>
    Shoes
  </div>
  <div>
    Too expensive
  </div>
  <div>
    450.00 z≈Ç
  </div>
  <div>
    2020-06-14
  </div>
  <div>
    <Link
      to="/edit/1"
    >
      Edit
    </Link>
  </div>
</div>
`;
